File: examples/footguns/02_data_race/main.c
/*
 * Footgun: Data Race — C
 *
 * Two threads increment a shared counter 1,000,000 times each.
 * Expected result: 2,000,000. Actual result: less, unpredictably.
 *
 * The read-increment-write of `counter++` is three instructions.
 * Both threads can read the same value, both increment it, and both
 * write back — losing one of the increments. C has no concept of
 * data-race-free by default; this compiles and runs silently wrong.
 */

#include <pthread.h>
#include <stdio.h>

#define ITERS 1000000L

static long counter = 0;

static void *increment(void *arg) {
    (void)arg;
    for (long i = 0; i < ITERS; i++)
        counter++; /* not atomic: load / add / store race */
    return NULL;
}

int main(void) {
    pthread_t t1, t2;
    pthread_create(&t1, NULL, increment, NULL);
    pthread_create(&t2, NULL, increment, NULL);
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    /* Almost never prints 2000000. */
    printf("counter = %ld (expected %ld)\n", counter, 2 * ITERS);
    return 0;
}

/*
 * Compile and run:
 *   gcc -pthread -O2 -o race main.c && ./race
 *   gcc -pthread -O2 -fsanitize=thread -o race main.c && ./race  # TSan
 *
 * Result without sanitizers: wrong answer, no diagnostic.
 * Result with TSan: WARNING: ThreadSanitizer: data race
 *
 * Fix: use _Atomic long counter, or protect with pthread_mutex_t.
 */

File: examples/footguns/02_data_race/main.go
// Footgun: Data Race — Go
//
// Go detects data races, but only at runtime and only with -race.
// Without that flag the program silently produces wrong output.
// The race detector is opt-in, not guaranteed, and has ~5-20x overhead
// so it is rarely enabled in production.

package main

import (
	"fmt"
	"sync"
)

const iters = 1_000_000

func main() {
	var counter int // shared, unprotected
	var wg sync.WaitGroup
	wg.Add(2)

	go func() {
		defer wg.Done()
		for i := 0; i < iters; i++ {
			counter++ // data race: read-modify-write is not atomic
		}
	}()

	go func() {
		defer wg.Done()
		for i := 0; i < iters; i++ {
			counter++ // data race
		}
	}()

	wg.Wait()
	fmt.Printf("counter = %d (expected %d)\n", counter, 2*iters)
}

// Compile and run:
//   go run main.go              # wrong answer, no diagnostic
//   go run -race main.go        # WARNING: DATA RACE (correct detection)
//
// Fix: use sync/atomic or protect with sync.Mutex.
//
// Note: Go's memory model defines data races as undefined behavior,
// but the toolchain does not prevent them — it only detects them
// if you ask it to.
